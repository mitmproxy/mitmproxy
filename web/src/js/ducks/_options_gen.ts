/** Auto-generated by test_app.py:test_generate_options_js */
export interface OptionsState {
    add_upstream_certs_to_client_chain: boolean
    allow_hosts: string[]
    anticache: boolean
    anticomp: boolean
    block_global: boolean
    block_list: string[]
    block_private: boolean
    body_size_limit: string | undefined
    cert_passphrase: string | undefined
    certs: string[]
    ciphers_client: string | undefined
    ciphers_server: string | undefined
    client_certs: string | undefined
    client_replay: string[]
    client_replay_concurrency: number
    command_history: boolean
    confdir: string
    connect_addr: string | undefined
    connection_strategy: string
    console_focus_follow: boolean
    content_view_lines_cutoff: number
    dns_listen_host: string
    dns_listen_port: number
    dns_mode: string
    dns_server: boolean
    export_preserve_original_ip: boolean
    http2: boolean
    http2_ping_keepalive: number
    ignore_hosts: string[]
    intercept: string | undefined
    intercept_active: boolean
    keep_host_header: boolean
    key_size: number
    listen_host: string
    listen_port: number
    map_local: string[]
    map_remote: string[]
    mode: string
    modify_body: string[]
    modify_headers: string[]
    normalize_outbound_headers: boolean
    onboarding: boolean
    onboarding_host: string
    onboarding_port: number
    proxy_debug: boolean
    proxyauth: string | undefined
    rawtcp: boolean
    readfile_filter: string | undefined
    rfile: string | undefined
    save_stream_file: string | undefined
    save_stream_filter: string | undefined
    scripts: string[]
    server: boolean
    server_replay: string[]
    server_replay_ignore_content: boolean
    server_replay_ignore_host: boolean
    server_replay_ignore_params: string[]
    server_replay_ignore_payload_params: string[]
    server_replay_ignore_port: boolean
    server_replay_kill_extra: boolean
    server_replay_nopop: boolean
    server_replay_refresh: boolean
    server_replay_use_headers: string[]
    showhost: boolean
    ssl_insecure: boolean
    ssl_verify_upstream_trusted_ca: string | undefined
    ssl_verify_upstream_trusted_confdir: string | undefined
    stickyauth: string | undefined
    stickycookie: string | undefined
    stream_large_bodies: string | undefined
    tcp_hosts: string[]
    termlog_verbosity: string
    tls_version_client_max: string
    tls_version_client_min: string
    tls_version_server_max: string
    tls_version_server_min: string
    upstream_auth: string | undefined
    upstream_cert: boolean
    validate_inbound_headers: boolean
    view_filter: string | undefined
    view_order: string
    view_order_reversed: boolean
    web_columns: string[]
    web_debug: boolean
    web_host: string
    web_open_browser: boolean
    web_port: number
    web_static_viewer: string | undefined
    websocket: boolean
}

export type Option = keyof OptionsState

export const defaultState: OptionsState = {
    add_upstream_certs_to_client_chain: false,
    allow_hosts: [],
    anticache: false,
    anticomp: false,
    block_global: true,
    block_list: [],
    block_private: false,
    body_size_limit: undefined,
    cert_passphrase: undefined,
    certs: [],
    ciphers_client: undefined,
    ciphers_server: undefined,
    client_certs: undefined,
    client_replay: [],
    client_replay_concurrency: 1,
    command_history: true,
    confdir: "~/.mitmproxy",
    connect_addr: undefined,
    connection_strategy: "eager",
    console_focus_follow: false,
    content_view_lines_cutoff: 512,
    dns_listen_host: "",
    dns_listen_port: 53,
    dns_mode: "regular",
    dns_server: false,
    export_preserve_original_ip: false,
    http2: true,
    http2_ping_keepalive: 58,
    ignore_hosts: [],
    intercept: undefined,
    intercept_active: false,
    keep_host_header: false,
    key_size: 2048,
    listen_host: "",
    listen_port: 8080,
    map_local: [],
    map_remote: [],
    mode: "regular",
    modify_body: [],
    modify_headers: [],
    normalize_outbound_headers: true,
    onboarding: true,
    onboarding_host: "mitm.it",
    onboarding_port: 80,
    proxy_debug: false,
    proxyauth: undefined,
    rawtcp: true,
    readfile_filter: undefined,
    rfile: undefined,
    save_stream_file: undefined,
    save_stream_filter: undefined,
    scripts: [],
    server: true,
    server_replay: [],
    server_replay_ignore_content: false,
    server_replay_ignore_host: false,
    server_replay_ignore_params: [],
    server_replay_ignore_payload_params: [],
    server_replay_ignore_port: false,
    server_replay_kill_extra: false,
    server_replay_nopop: false,
    server_replay_refresh: true,
    server_replay_use_headers: [],
    showhost: false,
    ssl_insecure: false,
    ssl_verify_upstream_trusted_ca: undefined,
    ssl_verify_upstream_trusted_confdir: undefined,
    stickyauth: undefined,
    stickycookie: undefined,
    stream_large_bodies: undefined,
    tcp_hosts: [],
    termlog_verbosity: "info",
    tls_version_client_max: "UNBOUNDED",
    tls_version_client_min: "TLS1_2",
    tls_version_server_max: "UNBOUNDED",
    tls_version_server_min: "TLS1_2",
    upstream_auth: undefined,
    upstream_cert: true,
    validate_inbound_headers: true,
    view_filter: undefined,
    view_order: "time",
    view_order_reversed: false,
    web_columns: ["tls", "icon", "path", "method", "status", "size", "time"],
    web_debug: false,
    web_host: "127.0.0.1",
    web_open_browser: true,
    web_port: 8081,
    web_static_viewer: "",
    websocket: true,
}
