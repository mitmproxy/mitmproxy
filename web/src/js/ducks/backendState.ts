/**
 * serverState houses properties about the current mitmproxy instance that are not options,
 * e.g. the list of available content views or the current version.
 */

export const RECEIVE = 'STATE_RECEIVE'
export const UPDATE = 'STATE_UPDATE'

export interface ServerInfo {
    description: string
    full_spec: string
    is_running: boolean
    last_exception: string | null
    listen_addrs: [string, number][]
    type: string
    wireguard_conf?: string
}

export interface BackendState {
    available: boolean
    version: string
    contentViews: string[]
    servers: ServerInfo[]
}

export const defaultState: BackendState = {
    available: false,
    version: "",
    contentViews: [],
    servers: [],
};

export default function reducer(state = defaultState, action): BackendState {
    switch (action.type) {
        case RECEIVE:
        case UPDATE:
            return {
                ...state,
                available: true,
                ...action.data
            }
        default:
            return state
    }
}
