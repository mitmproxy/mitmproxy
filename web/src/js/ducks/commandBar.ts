export const TOGGLE_VISIBILITY = "COMMANDBAR_TOGGLE_VISIBILITY";

interface CommandBarState {
    visible: boolean;
}

export const defaultState: CommandBarState = {
    visible: false,
};

export default function reducer(state = defaultState, action): CommandBarState {
    switch (action.type) {
        case TOGGLE_VISIBILITY:
            return {
                ...state,
                visible: !state.visible,
            };

        default:
            return state;
    }
}

export function toggleVisibility() {
    return { type: TOGGLE_VISIBILITY };
}
