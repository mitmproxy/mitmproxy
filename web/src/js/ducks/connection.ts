export enum ConnectionState {
    INIT = "CONNECTION_INIT",
    FETCHING = "CONNECTION_FETCHING", // WebSocket is established, but still fetching resources.
    ESTABLISHED = "CONNECTION_ESTABLISHED",
    ERROR = "CONNECTION_ERROR",
    OFFLINE = "CONNECTION_OFFLINE", // indicates that there is no live (websocket) backend.
}

interface ConnState {
    state: ConnectionState;
    message?: string;
}

const defaultState: ConnState = {
    state: ConnectionState.INIT,
    message: undefined,
};

export default function reducer(state = defaultState, action): ConnState {
    switch (action.type) {
        case ConnectionState.ESTABLISHED:
        case ConnectionState.FETCHING:
        case ConnectionState.ERROR:
        case ConnectionState.OFFLINE:
            return {
                state: action.type,
                message: action.message,
            };

        default:
            return state;
    }
}

export function startFetching() {
    return { type: ConnectionState.FETCHING };
}

export function connectionEstablished() {
    return { type: ConnectionState.ESTABLISHED };
}

export function connectionError(message) {
    return { type: ConnectionState.ERROR, message };
}

export function setOffline() {
    return { type: ConnectionState.OFFLINE };
}
