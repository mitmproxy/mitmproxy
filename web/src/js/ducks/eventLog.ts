import * as store from "./utils/store";

export const ADD = "EVENTS_ADD";
export const RECEIVE = "EVENTS_RECEIVE";
export const TOGGLE_VISIBILITY = "EVENTS_TOGGLE_VISIBILITY";
export const TOGGLE_FILTER = "EVENTS_TOGGLE_FILTER";

type LogLevel = "debug" | "info" | "web" | "warn" | "error";

export interface EventLogItem extends store.Item {
    message: string;
    level: LogLevel;
}

interface EventLogState extends store.State<EventLogItem> {
    visible: boolean;
    filters: { [level in LogLevel]: boolean };
}

const defaultState: EventLogState = {
    visible: false,
    filters: { debug: false, info: true, web: true, warn: true, error: true },
    ...store.defaultState,
};

export default function reduce(
    state: EventLogState = defaultState,
    action
): EventLogState {
    switch (action.type) {
        case TOGGLE_VISIBILITY:
            return {
                ...state,
                visible: !state.visible,
            };

        case TOGGLE_FILTER:
            const filters = {
                ...state.filters,
                [action.filter]: !state.filters[action.filter],
            };
            return {
                ...state,
                filters,
                ...store.reduce(
                    state,
                    store.setFilter<EventLogItem>((log) => filters[log.level])
                ),
            };

        case ADD:
        case RECEIVE:
            return {
                ...state,
                ...store.reduce(
                    state,
                    store[action.cmd](
                        action.data,
                        (log) => state.filters[log.level]
                    )
                ),
            };

        default:
            return state;
    }
}

export function toggleFilter(filter: LogLevel) {
    return { type: TOGGLE_FILTER, filter };
}

export function toggleVisibility() {
    return { type: TOGGLE_VISIBILITY };
}

export function add(message: string, level: LogLevel = "web") {
    let data = {
        id: Math.random().toString(),
        message,
        level,
    };
    return {
        type: ADD,
        cmd: "add",
        data,
    };
}
