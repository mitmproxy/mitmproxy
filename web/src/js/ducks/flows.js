import { fetchApi } from "../utils"
import reduceStore from "./utils/store"
import * as storeActions from "./utils/store"
import Filt from "../filt/filt"
import { RequestUtils } from "../flow/utils"

export const ADD            = 'FLOWS_ADD'
export const UPDATE         = 'FLOWS_UPDATE'
export const REMOVE         = 'FLOWS_REMOVE'
export const RECEIVE        = 'FLOWS_RECEIVE'
export const SELECT         = 'FLOWS_SELECT'
export const SET_FILTER     = 'FLOWS_SET_FILTER'
export const SET_SORT       = 'FLOWS_SET_SORT'
export const SET_HIGHLIGHT  = 'FLOWS_SET_HIGHLIGHT'
export const REQUEST_ACTION = 'FLOWS_REQUEST_ACTION'


const defaultState = {
    highlight: null,
    filter: null,
    sort: { column: null, desc: false },
    selected: [],
    ...reduceStore(undefined, {})
}

export default function reduce(state = defaultState, action) {
    switch (action.type) {

        case ADD:
        case UPDATE:
        case REMOVE:
        case RECEIVE:
            let storeAction = storeActions[action.cmd](
                action.data,
                makeFilter(state.filter),
                makeSort(state.sort)
            )

            let selected = state.selected
            if(action.type === REMOVE && state.selected.includes(action.data)) {
                if(state.selected.length > 1){
                    selected = selected.filter(x => x !== action.data)
                } else {
                    selected = []
                    if (action.data in state.viewIndex && state.view.length > 1) {
                        let currentIndex = state.viewIndex[action.data],
                            nextSelection
                        if(currentIndex === state.view.length -1){ // last row
                            nextSelection = state.view[currentIndex - 1]
                        } else {
                            nextSelection = state.view[currentIndex + 1]
                        }
                        selected.push(nextSelection.id)
                    }
                }
            }

            return {
                ...state,
                selected,
                ...reduceStore(state, storeAction)
            }

        case SET_FILTER:
            return {
                ...state,
                filter: action.filter,
                ...reduceStore(state, storeActions.setFilter(makeFilter(action.filter), makeSort(state.sort)))
            }

        case SET_HIGHLIGHT:
            return {
                ...state,
                highlight: action.highlight
            }

        case SET_SORT:
            return {
                ...state,
                sort: action.sort,
                ...reduceStore(state, storeActions.setSort(makeSort(action.sort)))
            }

        case SELECT:
            return {
                ...state,
                selected: action.flowIds
            }

        default:
            return state
    }
}


const sortKeyFuns = {

    TLSColumn: flow => flow.request.scheme,

    PathColumn: flow => RequestUtils.pretty_url(flow.request),

    MethodColumn: flow => flow.request.method,

    StatusColumn: flow => flow.response && flow.response.status_code,

    TimeColumn: flow => flow.response && flow.response.timestamp_end - flow.request.timestamp_start,

    SizeColumn: flow => {
        let total = flow.request.contentLength
        if (flow.response) {
            total += flow.response.contentLength || 0
        }
        return total
    },
}

export function makeFilter(filter) {
    if (!filter) {
        return
    }
    return Filt.parse(filter)
}

export function makeSort({ column, desc }) {
    const sortKeyFun = sortKeyFuns[column]
    if (!sortKeyFun) {
        return
    }
    return (a, b) => {
        const ka = sortKeyFun(a)
        const kb = sortKeyFun(b)
        if (ka > kb) {
            return desc ? -1 : 1
        }
        if (ka < kb) {
            return desc ? 1 : -1
        }
        return 0
    }
}

export function setFilter(filter) {
    return { type: SET_FILTER, filter }
}

export function setHighlight(highlight) {
    return { type: SET_HIGHLIGHT, highlight }
}

export function setSort(column, desc) {
    return { type: SET_SORT, sort: { column, desc } }
}

export function selectRelative(flows, shift) {
    let currentSelectionIndex = flows.viewIndex[flows.selected[0]]
    let minIndex = 0
    let maxIndex = flows.view.length - 1
    let newIndex
    if (currentSelectionIndex === undefined) {
        newIndex = (shift < 0) ? minIndex : maxIndex
    } else {
        newIndex = currentSelectionIndex + shift
        newIndex = window.Math.max(newIndex, minIndex)
        newIndex = window.Math.min(newIndex, maxIndex)
    }
    let flow = flows.view[newIndex]
    return select(flow ? flow.id : undefined)
}


export function resume(flow) {
    return dispatch => fetchApi(`/flows/${flow.id}/resume`, { method: 'POST' })
}

export function resumeAll() {
    return dispatch => fetchApi('/flows/resume', { method: 'POST' })
}

export function kill(flow) {
    return dispatch => fetchApi(`/flows/${flow.id}/kill`, { method: 'POST' })
}

export function killAll() {
    return dispatch => fetchApi('/flows/kill', { method: 'POST' })
}


export function remove(flow) {
    return dispatch => fetchApi(`/flows/${flow.id}`, { method: 'DELETE' })
}

export function duplicate(flow) {
    return dispatch => fetchApi(`/flows/${flow.id}/duplicate`, { method: 'POST' })
}

export function replay(flow) {
    return dispatch => fetchApi(`/flows/${flow.id}/replay`, { method: 'POST' })
}

export function revert(flow) {
    return dispatch => fetchApi(`/flows/${flow.id}/revert`, { method: 'POST' })
}

export function update(flow, data) {
    return dispatch => fetchApi.put(`/flows/${flow.id}`, data)
}

export function uploadContent(flow, file, type) {
    const body = new FormData()
    file       = new window.Blob([file], { type: 'plain/text' })
    body.append('file', file)
    return dispatch => fetchApi(`/flows/${flow.id}/${type}/content.data`, { method: 'POST', body })
}


export function clear() {
    return dispatch => fetchApi('/clear', { method: 'POST' })
}

export function download() {
    window.location = '/flows/dump'
    return { type: REQUEST_ACTION }
}

export function upload(file) {
    const body = new FormData()
    body.append('file', file)
    return dispatch => fetchApi('/flows/dump', { method: 'POST', body })
}


export function select(id) {
    return {
        type: SELECT,
        flowIds: id ? [id] : []
    }
}
