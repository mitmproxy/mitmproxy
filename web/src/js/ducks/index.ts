import {
    AnyAction,
    applyMiddleware,
    combineReducers,
    compose,
    createStore as createReduxStore,
    PreloadedState,
    Store
} from "redux"
import eventLog from "./eventLog"
import flows from "./flows"
import ui from "./ui/index"
import connection from "./connection"
import options from './options'
import commandBar from "./commandBar";
import thunk, {ThunkAction, ThunkDispatch, ThunkMiddleware} from "redux-thunk";
import {TypedUseSelectorHook, useDispatch, useSelector} from "react-redux";
import backendState from "./backendState";
import options_meta from "./options_meta";


// @ts-ignore
const composeEnhancers = window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose;

export const rootReducer = combineReducers({
    commandBar,
    eventLog,
    flows,
    connection,
    ui,
    options,
    options_meta,
    backendState,
});
export type RootState = ReturnType<typeof rootReducer>


export type AppThunk<ReturnType = void> = ThunkAction<ReturnType,
    RootState,
    unknown,
    AnyAction>

export const createAppStore = (preloadedState?: PreloadedState<RootState>): Store<RootState> => {
    return createReduxStore(
        rootReducer,
        preloadedState,
        composeEnhancers(applyMiddleware(
            thunk as ThunkMiddleware<RootState>
        )))
};

export const store = createAppStore(undefined);

// this would be correct, but PyCharm bails on it
// export type AppDispatch = typeof store.dispatch
// instead:
export type AppDispatch = ThunkDispatch<RootState, void, AnyAction>
export const useAppDispatch = () => useDispatch<AppDispatch>()
export const useAppSelector: TypedUseSelectorHook<RootState> = useSelector
