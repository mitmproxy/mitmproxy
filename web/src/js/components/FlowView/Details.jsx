import React from 'react'
import _ from 'lodash'
import { formatTimeStamp, formatTimeDelta } from '../../utils.js'

export function TimeStamp({ t, deltaTo, title }) {
    return t ? (
        <tr>
            <td>{title}:</td>
            <td>
                {formatTimeStamp(t)}
                {deltaTo && (
                    <span className="text-muted">
                        ({formatTimeDelta(1000 * (t - deltaTo))})
                    </span>
                )}
            </td>
        </tr>
    ) : (
        <tr></tr>
    )
}

export function ConnectionInfo({ conn }) {    
    return (
        <table className="connection-table">
            <tbody>
                    <tr key="address">
                    <td>Address:</td>
                    <td>{conn.address.join(':')}</td>
                </tr>      
                {conn.sni && (
                    <tr key="sni">
                        <td><abbr title="TLS Server Name Indication">TLS SNI:</abbr></td>
                        <td>{conn.sni}</td>
                    </tr>
                )}
                {conn.tls_version && (
                    <tr key="tls_version">
                        <td>TLS version:</td>
                        <td>{conn.tls_version}</td>
                    </tr>
                )}
                {conn.cipher_name && (
                    <tr key="cipher_name">
                        <td>cipher name:</td>
                        <td>{conn.cipher_name}</td>
                    </tr>
                )}
                {conn.alpn_proto_negotiated && (
                    <tr key="ALPN">
                        <td><abbr title="ALPN protocol negotiated">ALPN:</abbr></td>
                        <td>{conn.alpn_proto_negotiated}</td>
                    </tr>
                )}
            </tbody>
        </table>
    )
}

export function CertificateInfo({ flow }) {
    // @todo We should fetch human-readable certificate representation from the server
    return (
        <div>
            {flow.client_conn.cert && [
                <h4 key="name">Client Certificate</h4>,
                <pre key="value" style={{ maxHeight: 100 }}>{flow.client_conn.cert}</pre>
            ]}

            {flow.server_conn.cert && [
                <h4 key="name">Server Certificate</h4>,
                <pre key="value" style={{ maxHeight: 100 }}>{flow.server_conn.cert}</pre>
            ]}
        </div>
    )
}

export function Timing({ flow }) {
    const { server_conn: sc, client_conn: cc, request: req, response: res } = flow

    const timestamps = [
        {
            title: "Server conn. initiated",
            t: sc.timestamp_start,
            deltaTo: req.timestamp_start
        }, {
            title: "Server conn. TCP handshake",
            t: sc.timestamp_tcp_setup,
            deltaTo: req.timestamp_start
        }, {
            title: "Server conn. SSL handshake",
            t: sc.timestamp_ssl_setup,
            deltaTo: req.timestamp_start
        }, {
            title: "Client conn. established",
            t: cc.timestamp_start,
            deltaTo: req.timestamp_start
        }, {
            title: "Client conn. SSL handshake",
            t: cc.timestamp_ssl_setup,
            deltaTo: req.timestamp_start
        }, {
            title: "First request byte",
            t: req.timestamp_start
        }, {
            title: "Request complete",
            t: req.timestamp_end,
            deltaTo: req.timestamp_start
        }, res && {
            title: "First response byte",
            t: res.timestamp_start,
            deltaTo: req.timestamp_start
        }, res && {
            title: "Response complete",
            t: res.timestamp_end,
            deltaTo: req.timestamp_start
        }
    ]

    return (
        <div>
            <h4>Timing</h4>
            <table className="timing-table">
                <tbody>
                    {timestamps.filter(v => v).sort((a, b) => a.t - b.t).map(item => (
                        <TimeStamp key={item.title} {...item}/>
                    ))}
                </tbody>
            </table>
        </div>
    )
}

export default function Details({ flow }) {
    return (
        <section className="detail">
            <h4>Client Connection</h4>
            <ConnectionInfo conn={flow.client_conn}/>

            {flow.server_conn.address && 
                    [
                        <h4>Server Connection</h4>,
                        <ConnectionInfo conn={flow.server_conn}/>
                    ]    
            }
            
            <CertificateInfo flow={flow}/>

            <Timing flow={flow}/>
        </section>
    )
}
