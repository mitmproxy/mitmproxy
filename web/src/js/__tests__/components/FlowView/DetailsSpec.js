import React from 'react'
import renderer from 'react-test-renderer'
import Details, { TimeStamp, ConnectionInfo, CertificateInfo, Timing } from '../../../components/FlowView/Details'
import { TFlow } from '../../ducks/tutils'

let tflow = TFlow()

describe('TimeStamp Component', () => {
    it('should render correctly', () => {
        let timestamp = renderer.create(<TimeStamp t={1483228800} deltaTo={1483228700} title="foo"/>),
            tree = timestamp.toJSON()
        expect(tree).toMatchSnapshot()
        // without timestamp
        timestamp = renderer.create(<TimeStamp/>)
        tree = timestamp.toJSON()
        expect(tree).toMatchSnapshot()
    })
})

describe('ConnectionInfo Component', () => {
    it('should render correctly', () => {
        let connectionInfo = renderer.create(<ConnectionInfo conn={tflow.client_conn}/>),
            tree = connectionInfo.toJSON()
        expect(tree).toMatchSnapshot()
    })
})

describe('CertificateInfo Component', () => {
    it('should render correctly', () => {
        let certificateInfo = renderer.create(<CertificateInfo flow={tflow}/>),
            tree = certificateInfo.toJSON()
        expect(tree).toMatchSnapshot()
    })
})

describe('Timing Component', () => {
    it('should render correctly', () => {
        let timing = renderer.create(<Timing flow={tflow}/>),
            tree = timing.toJSON()
        expect(tree).toMatchSnapshot()
    })
})

describe('Details Component', () => {
    it('should render correctly', () => {
        let details = renderer.create(<Details flow={tflow}/>),
            tree = details.toJSON()
        expect(tree).toMatchSnapshot()
    })

    it('should render correctly when server address is missing', () => {
        let tflowServerAddressNull = tflow

        tflowServerAddressNull.server_conn.address = null
        tflowServerAddressNull.server_conn.ip_address = null
        tflowServerAddressNull.server_conn.alpn_proto_negotiated = null
        tflowServerAddressNull.server_conn.sni = null
        tflowServerAddressNull.server_conn.ssl_established = false
        tflowServerAddressNull.server_conn.tls_version = null
        tflowServerAddressNull.server_conn.timestamp_tcp_setup = null
        tflowServerAddressNull.server_conn.timestamp_ssl_setup = null
        tflowServerAddressNull.server_conn.timestamp_start = null
        tflowServerAddressNull.server_conn.timestamp_end = null
        
        let details = renderer.create(<Details flow={tflowServerAddressNull}/>),
            tree = details.toJSON()
        expect(tree).toMatchSnapshot()
    })

})
