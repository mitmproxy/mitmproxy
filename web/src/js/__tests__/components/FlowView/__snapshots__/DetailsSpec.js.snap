// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CertificateInfo Component should render correctly 1`] = `<div />`;

exports[`ConnectionInfo Component should render correctly 1`] = `
<table
  className="connection-table"
>
  <tbody>
    <tr>
      <td>
        Address:
      </td>
      <td>
        127.0.0.1:22
      </td>
    </tr>
    <tr>
      <td>
        <abbr
          title="TLS Server Name Indication"
        >
          TLS SNI:
        </abbr>
      </td>
      <td>
        address
      </td>
    </tr>
    <tr>
      <td>
        TLS version:
      </td>
      <td>
        TLSv1.2
      </td>
    </tr>
    <tr>
      <td>
        cipher name:
      </td>
      <td>
        cipher
      </td>
    </tr>
    <tr>
      <td>
        <abbr
          title="ALPN protocol negotiated"
        >
          ALPN:
        </abbr>
      </td>
      <td>
        http/1.1
      </td>
    </tr>
  </tbody>
</table>
`;

exports[`Details Component should render correctly 1`] = `
<section
  className="detail"
>
  <h4>
    Client Connection
  </h4>
  <table
    className="connection-table"
  >
    <tbody>
      <tr>
        <td>
          Address:
        </td>
        <td>
          127.0.0.1:22
        </td>
      </tr>
      <tr>
        <td>
          <abbr
            title="TLS Server Name Indication"
          >
            TLS SNI:
          </abbr>
        </td>
        <td>
          address
        </td>
      </tr>
      <tr>
        <td>
          TLS version:
        </td>
        <td>
          TLSv1.2
        </td>
      </tr>
      <tr>
        <td>
          cipher name:
        </td>
        <td>
          cipher
        </td>
      </tr>
      <tr>
        <td>
          <abbr
            title="ALPN protocol negotiated"
          >
            ALPN:
          </abbr>
        </td>
        <td>
          http/1.1
        </td>
      </tr>
    </tbody>
  </table>
  <h4>
    Server Connection
  </h4>
  <table
    className="connection-table"
  >
    <tbody>
      <tr>
        <td>
          Address:
        </td>
        <td>
          address:22
        </td>
      </tr>
      <tr>
        <td>
          <abbr
            title="TLS Server Name Indication"
          >
            TLS SNI:
          </abbr>
        </td>
        <td>
          address
        </td>
      </tr>
      <tr>
        <td>
          TLS version:
        </td>
        <td>
          TLSv1.2
        </td>
      </tr>
    </tbody>
  </table>
  <div />
  <div>
    <h4>
      Timing
    </h4>
    <table
      className="timing-table"
    >
      <tbody>
        <tr />
        <tr />
        <tr>
          <td>
            Server conn. initiated
            :
          </td>
          <td>
            1970-01-01 00:00:01.000
          </td>
        </tr>
        <tr>
          <td>
            Client conn. established
            :
          </td>
          <td>
            1970-01-01 00:00:01.000
          </td>
        </tr>
        <tr>
          <td>
            Server conn. TCP handshake
            :
          </td>
          <td>
            1970-01-01 00:00:02.000
          </td>
        </tr>
        <tr>
          <td>
            Client conn. SSL handshake
            :
          </td>
          <td>
            1970-01-01 00:00:02.000
          </td>
        </tr>
        <tr>
          <td>
            Server conn. SSL handshake
            :
          </td>
          <td>
            1970-01-01 00:00:03.000
          </td>
        </tr>
        <tr>
          <td>
            First response byte
            :
          </td>
          <td>
            2017-05-21 12:38:32.481
          </td>
        </tr>
        <tr>
          <td>
            Response complete
            :
          </td>
          <td>
            2017-05-21 12:38:32.481
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</section>
`;

exports[`Details Component should render correctly when server address is missing 1`] = `
<section
  className="detail"
>
  <h4>
    Client Connection
  </h4>
  <table
    className="connection-table"
  >
    <tbody>
      <tr>
        <td>
          Address:
        </td>
        <td>
          127.0.0.1:22
        </td>
      </tr>
      <tr>
        <td>
          <abbr
            title="TLS Server Name Indication"
          >
            TLS SNI:
          </abbr>
        </td>
        <td>
          address
        </td>
      </tr>
      <tr>
        <td>
          TLS version:
        </td>
        <td>
          TLSv1.2
        </td>
      </tr>
      <tr>
        <td>
          cipher name:
        </td>
        <td>
          cipher
        </td>
      </tr>
      <tr>
        <td>
          <abbr
            title="ALPN protocol negotiated"
          >
            ALPN:
          </abbr>
        </td>
        <td>
          http/1.1
        </td>
      </tr>
    </tbody>
  </table>
  <div />
  <div>
    <h4>
      Timing
    </h4>
    <table
      className="timing-table"
    >
      <tbody>
        <tr />
        <tr />
        <tr />
        <tr />
        <tr />
        <tr>
          <td>
            Client conn. established
            :
          </td>
          <td>
            1970-01-01 00:00:01.000
          </td>
        </tr>
        <tr>
          <td>
            Client conn. SSL handshake
            :
          </td>
          <td>
            1970-01-01 00:00:02.000
          </td>
        </tr>
        <tr>
          <td>
            First response byte
            :
          </td>
          <td>
            2017-05-21 12:38:32.481
          </td>
        </tr>
        <tr>
          <td>
            Response complete
            :
          </td>
          <td>
            2017-05-21 12:38:32.481
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</section>
`;

exports[`TimeStamp Component should render correctly 1`] = `
<tr>
  <td>
    foo
    :
  </td>
  <td>
    2017-01-01 00:00:00.000
    <span
      className="text-muted"
    >
      (
      2min
      )
    </span>
  </td>
</tr>
`;

exports[`TimeStamp Component should render correctly 2`] = `<tr />`;

exports[`Timing Component should render correctly 1`] = `
<div>
  <h4>
    Timing
  </h4>
  <table
    className="timing-table"
  >
    <tbody>
      <tr />
      <tr />
      <tr>
        <td>
          Server conn. initiated
          :
        </td>
        <td>
          1970-01-01 00:00:01.000
        </td>
      </tr>
      <tr>
        <td>
          Client conn. established
          :
        </td>
        <td>
          1970-01-01 00:00:01.000
        </td>
      </tr>
      <tr>
        <td>
          Server conn. TCP handshake
          :
        </td>
        <td>
          1970-01-01 00:00:02.000
        </td>
      </tr>
      <tr>
        <td>
          Client conn. SSL handshake
          :
        </td>
        <td>
          1970-01-01 00:00:02.000
        </td>
      </tr>
      <tr>
        <td>
          Server conn. SSL handshake
          :
        </td>
        <td>
          1970-01-01 00:00:03.000
        </td>
      </tr>
      <tr>
        <td>
          First response byte
          :
        </td>
        <td>
          2017-05-21 12:38:32.481
        </td>
      </tr>
      <tr>
        <td>
          Response complete
          :
        </td>
        <td>
          2017-05-21 12:38:32.481
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
